src/System.Diagnostics.TraceSource/src/System/Diagnostics/CorrelationManager.cs

```diff
@@ -0,0 +1,72 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Threading;
+
+namespace System.Diagnostics
+{
+    public class CorrelationManager
+    {
+        internal CorrelationManager()
+        {
+        }
+
+        private readonly AsyncLocal<Stack> _idStack = new AsyncLocal<Stack>();
```

Comment:
Unfortunately as I mentioned offline, I think both (1) and (2) have issues.  For (1), we can't change the public signature, and if we just return a new Stack from the property, we lose updates developers make to it directly through the Stack.  For (2), the cloning would happen on another thread, which means we could be reading from it while the original thread is mutating it.

Here's my idea.  It's a tweak on (1) that takes advantage of the fact that Stack isn't sealed and all of its members are virtual!  So it essentially does (1), but returns a Stack that wraps the underlying immutable stack (just a simple linked list, no need to bring in the immutable collections library), and then all updates (internal and external) are made via that derived Stack.  We actually propagate the immutable stack, though, and thus don't need to worry about concurrent mutation.

```C#
namespace System.Diagnostics
{
    public class CorrelationManager
    {
        private readonly AsyncLocal<Guid> _activityId = new AsyncLocal<Guid>();
        private readonly AsyncLocal<StackNode> _stack = new AsyncLocal<StackNode>();
        private readonly Stack _stackWrapper;

        internal CorrelationManager()
        {
            _stackWrapper = new AsyncLocalStackWrapper(_stack);
        }

        public Guid ActivityId { get { return _activityId.Value; } set { _activityId.Value = value; } }

        public Stack LogicalOperationStack => _stackWrapper;

        public void StartLogicalOperation() => StartLogicalOperation(Guid.NewGuid());

        public void StartLogicalOperation(object operationId)
        {
            if (operationId == null)
            {
                throw new ArgumentNullException(nameof(operationId));
            }

            _stackWrapper.Push(operationId);
        }

        public void StopLogicalOperation() => _stackWrapper.Pop();

        private sealed class StackNode
        {
            internal StackNode(object value, StackNode prev = null)
            {
                Value = value;
                Prev = prev;
                Count = prev != null ? prev.Count + 1 : 1;
            }

            internal int Count { get; }
            internal object Value { get; }
            internal StackNode Prev { get; }
        }

        private sealed class AsyncLocalStackWrapper : Stack
        {
            private readonly AsyncLocal<StackNode> _stack;

            internal AsyncLocalStackWrapper(AsyncLocal<StackNode> stack)
            {
                Debug.Assert(stack != null);
                _stack = stack;
            }

            public override void Clear() => _stack.Value = null;

            public override object Clone() => new AsyncLocalStackWrapper(_stack);

            public override bool Contains(object obj)
            {
                for (StackNode n = _stack.Value; n != null; n = n.Prev)
                {
                    if (obj == null)
                    {
                        if (n.Value == null) return true;
                    }
                    else if (obj.Equals(n.Value))
                    {
                        return true;
                    }
                }
                return false;
            }

            public override void CopyTo(Array array, int index)
            {
                for (StackNode n = _stack.Value; n != null; n = n.Prev)
                {
                    array.SetValue(n.Value, index++);
                }
            }

            public override int Count => _stack.Value?.Count ?? 0;

            public override IEnumerator GetEnumerator() => GetEnumerator(_stack.Value);
            private IEnumerator GetEnumerator(StackNode n)
            {
                while (n != null)
                {
                    yield return n.Value;
                    n = n.Prev;
                }
            }

            public override object Peek() => _stack.Value?.Value;

            public override object Pop()
            {
                StackNode n = _stack.Value;
                if (n == null)
                {
                    throw new InvalidOperationException();
                }
                _stack.Value = n.Prev;
                return n.Value;
            }

            public override void Push(object obj)
            {
                _stack.Value = new StackNode(obj, _stack.Value);
            }

            public override object[] ToArray()
            {
                StackNode n = _stack.Value;
                if (n == null)
                {
                    return Array.Empty<object>();
                }

                var results = new List<object>();
                do
                {
                    results.Add(n.Value);
                    n = n.Prev;
                }
                while (n != null);
                return results.ToArray();
            }
        }
    }
}
```
Take this with a grain of salt... I've not tested it so there may be an issue here and there.

Thoughts?